prefetch.bs

```diff
@@ -0,0 +1,315 @@
+
+    1. [=Assert=]: |document| is [=Document/fully active=].
+    1. Let |currentTime| be the [=current high resolution time=] for the [=relevant global object=] of |document|.
+    1. Let |expiryTime| be |currentTime| + 300000 (i.e., five minutes).
+    1. [=list/Remove=] all elements whose [=prefetch record/URL=] equals |url| and [=prefetch record/referrer policy=] equals |referrerPolicy| from |document|'s [=prefetch buffer=].
+    1. [=list/Append=] a [=prefetch record=] with [=prefetch record/URL=] |url|, [=prefetch record/referrer policy=] |referrerPolicy|, [=prefetch record/response=] |response| and [=prefetch record/expiry time=] |expiryTime| to |document|'s [=prefetch buffer=].
+</div>
+
+<div algorithm="find a matching prefetch response">
+    To <dfn export>find a matching prefetch response</dfn> given a {{Document}} |document|, [=URL=] |url| and [=referrer policy=] |referrerPolicy|, perform the following steps.
+
+    1. [=Assert=]: |document| is [=Document/fully active=].
+    1. Let |currentTime| be the [=current high resolution time=] for the [=relevant global object=] of |document|.
+    1. [=list/For each=] |record| of |document|'s [=prefetch buffer=]:
+        1. If |record|'s [=prefetch record/URL=] is not equal to |url| or |record|'s [=prefetch record/referrer policy=] is not equal to |referrerPolicy|, then [=iteration/continue=].
+        1. [=list/Remove=] |record| from |document|'s [=prefetch buffer=].
+        1. If |record|'s [=prefetch record/expiry time=] is less than |currentTime|, return null.
+        1. Return |record|'s [=prefetch record/response=].
+    1. Return null.
+    
+    <p class="issue">It might be possible to use cache response headers to determine when a response can be used multiple times, but given the short lifetime of the prefetch buffer it's unclear whether this is worthwhile.</p>
+</div>
+
+<h2 id="html-patches">HTML Patches</h2>
+
+<div algorithm="perform a common navigational fetch">
+    <div class="note">This is an abstraction of the existing [=process a navigate fetch=]. It includes behavior like redirect handling that is particular to navigational fetches, including those that are relate to a speculate future navigation, rather than an immediate one.</div>
+
+    To <dfn export>perform a common navigational fetch</dfn> given a [=request=] |request|, [=string=] |navigationType|, [=browsing context=] |browsingContext|, [=environment=] or null |forceEnvironment|, algorithm |willRedirect| (which takes [=URLs=] currentURL and locationURL) algorithm |shouldBlockNavigationRequest| (which takes a [=request=], navigation type [=string=], and [=environment=], and returns "`Blocked`" or "`Allowed`"), algorithm |shouldBlockNavigationResponse| (which takes a [=request=] and a [=response=], and returns "`Blocked`" or "`Allowed`"), perform the following steps.
```

Comment:
"willRedirect" name is confusing; I thought it'd return a boolean. Maybe "processRedirect"?

Reply:
Hmm, the calling code still does the redirect processing (at least, I would interpret that to mean the updated URL computation and fetch). I was hoping to follow the common-ish (at least in code) `willDoFoo`/`didDooFoo` pattern for pre/post hooks. But if this is unclear (I can see how it has a boolean sense), how about `preRedirectHook`?

Reply:
Yeah, `preRedirectHook` seems good too.